File: Markov_Python/hmm.py
from copy import deepcopy

class HMM(object):
    def __init__(self, trans={}, state={}, emission={}):
        self.trans=trans
        self.state=state
        self.emission=emission

    def setTrans(self,t):
        if type(t)==type({}):
            self.trans=t
    
    def setTKey(self,tk=[]):
        if type(tk)==type([]):
            self.trans={}
            # set up rows
            for item in tk:
                self.trans.append({item: {}})
            # set up each row
            for k,v in self.trans.iteritems():
                v=map(lambda x: {x: 0.0}, tk)
    
    def setTProb(self, tv={}):
        # note: tv must have the form of
        # {key: {k : v}} where key, k, are string ,
        # v is float 
        if type(tv)==type({}):
            for k,v in tv:
                self.trans[k]=v
    
    def setEmit(self, e):
        if type(e)==type({}):
            self.emission = e
    
    def setEKey(self, ek):
        if type(ek)==type([]):
            self.trans={}
            # set up rows
            for item in ek[0]:
                self.emission.append({item: {}})
            # set up each row
            for k,v in self.emission.iteritems():
                v=map(lambda x: {x: 0.0}, ek[1])
    
    def setEProb(self, ev={}):
        # note: ev must have the form of
        # {key: {k : v}} where key, k, are string ,
        # v is float
        if type(ev)==type({}):
            for k,v in ev:
                self.emission[k]=v
    
    def setState(self, s):
        if type(s)==type({}):
            self.state=s
    
    def __forwardAlg(self, observation, state_prob, prob, index):
        module = {k: 0.0 for k, v in self.trans.iteritems()}
        for i in observation:
            # update the probability dictionary list
            # calculate the probability of the current observation

            # P(observe | state) = Sigma ( P(hidden) * P( hidden -> observe))
            # where hidden are keys in state dictionary, P( hidden -> observe)
            # is given by emission dictionary(matrix)
            # P(observe | state) = Matrix(state) * Matrix(eimission)
            p = 0.0
            for hidden in state_prob[index]:
                p += state_prob[index][hidden] * self.emission[hidden][i]
            prob *= p
            # calculate the next state

            # P(state) = initial state (when index == 0)
            # P(state) = Matrix(state-1) * Matrix(transition)
            for key, val in self.trans.iteritems():
                for k in module:
                    module[k]+=state_prob[index][key] * val[k]
            # append the new state to state list
            state_prob.append(deepcopy(module))
            # clear module
            module = {k: 0.0 for k, v in self.trans.iteritems()}
            index += 1
        return (prob, index)
    
    def evaluate(self, observation, target=[]):
        state_prob = [self.state]
        prob = 1.0
        index = 0

        # calculate the probability by forward algorithm
        prob, index = self.__forwardAlg(observation, state_prob, prob, index)

        # if we have target, it calculates
        # P( target_observation | existing observation )
        if target!=[]:
            prob_existing = prob
            prob, index = self.__forwardAlg(target, state_prob, prob, index)
            prob = prob / prob_existing 
        return prob
    
    def decode(self, observation):
        module = {k:0.0 for k, v in self.trans.iteritems()}
        state_prob = [deepcopy(module)]
        state_seq = []
        index = 0
        
        for i in observation:
            # update the probability dictionary list
            # at the beginning, only one observation
            if index == 0:
                # update the state_prob list 
                for k in state_prob[index]:
                    # state_prob[index][k]=P(k->i)
                    state_prob[index][k]=self.state[k]*self.emission[k][i]
            # if it is not the beginning 
            else:
                for k in state_prob[index]:
                    # state_prob[index][k]=P( k->i | last->k )*P(last)
                    last_hidden = state_seq[index-1]
                    last_prob = state_prob[index-1][last_hidden]
                    k_to_i = self.emission[k][i]
                    last_to_k = self.trans[last_hidden][k]
                    state_prob[index][k] = last_prob * last_to_k * k_to_i
            # update complete, finde the maximum prob and update it to sequnce
            maxed = False
            maxProb = 0.0
            for k in state_prob[index]:
                if not maxed:
                    maxed = True
                    maxProb=state_prob[index][k]
                    state_seq.append(k)
                elif state_prob[index][k] > maxProb:
                    maxProb=state_prob[index][k]
                    state_seq[index] = k
                else:
                    continue
            index+=1
            state_prob.append(deepcopy(module))
        return state_seq

            

# test code

h=HMM()
h.setTrans({"Healthy":{"Healthy":0.7, "Fever":0.3},
            "Fever":{"Healthy":0.4, "Fever": 0.6}})
h.setEmit({"Healthy":{"normal":0.5, "cold":0.4,"dizzy":0.1},
            "Fever":{"normal":0.1, "cold":0.3,"dizzy":0.6}})
h.setState({"Healthy": 0.6, "Fever": 0.4})

File: Markov_CPP/hmm.cc
#include "hmm.h"
#include <iostream>
using namespace std;

//private
/*
std::map<std::string, std::map<std::string,double>> trans;
std::map<std::string, std::map<std::string,double>> emission;
std::map<std::string, double> state;
std::map<std::string, double> module;
std::vector<std::string> hidden;
std::vector<std::string> shown;
*/

void HMM::forwardAlg(vector<string> &obs,
		vector<map<string, double>> &states,
		double &prob, int &index)
{
  double p;
  for(string i: obs){
    p=0.0;
    // P(shown | states) = Matrix(emission)[shown] * Matrix(state)
    for(auto j:states[index]){
      p+=j.second * emission[j.first][i];
    }
    prob*=p;
    index+=1;
    // P(state | state-1) = Matrix(state-1) * Matrix(trans)
    states.emplace_back(module);
    for(auto j:states[index]){
      for(auto k:trans[j.first]){
	states[index][k.first]+=states[index-1][j.first] * k.second;
      }
    }
  }
}

void HMM::viterbiAlg(vector<string> &obs,
		     vector<string> &store,
		     vector<map<string, double>> &states,
		     int &index)
{
  for(string i: obs){
    if(index){
      for(auto cur: states[index]){
	// in general 
	// p(current|observe)=*p(current|last)
	//                   *p(emission[current][observe])
	// where p(current|last)=p(trans[last][current])
	string last = store[index-1];
	double last_cur = trans[last][cur.first];
	double cur_obs = emission[cur.first][i];
	double last_prob = states[index-1][last];
	//cout<<last<<", trans: "<<last_cur<<" emit: "<<cur_obs<<" last prob "<<last_prob<<" current: "<<cur.first<<endl;
	states[index][cur.first] = last_cur * cur_obs * last_prob;
      }
    }else{
      // on the first observation
      // p( current | observe ) = p(current) * p(emission[current][observe])
      for(auto j: states[index]){
	states[index][j.first] *= emission[j.first][i];
      }
    }
    // find the maximum probability
    //cout<<states[index].size()<<endl;
    string maximum = states[index].begin()->first; 
    double max_prob = states[index].begin()->second;
    for(auto j: states[index]){
      if(j.second > max_prob){
	max_prob=j.second;
	maximum=j.first;
      }
    }
    store.emplace_back(maximum);
    states.emplace_back(module);
    index+=1;
  }
}

bool HMM::valid(){
  return trans.size() && emission.size() && state.size() && module.size() && shown.size() && hidden.size();
}

void HMM::clear(){
  trans.erase(trans.begin(),trans.end());
  emission.erase(emission.begin(),emission.end());
  state.erase(state.begin(),state.end());
  module.erase(module.begin(),module.end());
  hidden.erase(hidden.begin(),hidden.end());
  shown.erase(shown.begin(),shown.end());
}

void HMM::generateHS(){
  if(emission.size()){
    bool already=false;
    for(auto i: emission){
      hidden.emplace_back(i.first);
      if(!already){
	for(auto j: i.second){
	  shown.emplace_back(j.first);
	}
      }
    }
  }
}

void HMM::generateM(){
  if(trans.size()){
    if(!module.size()){
      module.erase(module.begin(),module.end());
    }
    for(auto i: trans){
      module[i.first]=0.0;
    }
  }
}

void HMM::generateTE(){
  if(!emission.size() && !trans.size()){
    // the emission block
    map<string,double> block;
    for(string i: shown){
      block[i]=0.0;
    }
    // the trans block
    map<string, double> block1;
    for(string i: hidden){
      block[i]=0.0;
      // set the emission
      emission[i]=block;
    }
    // set the trans
    for(string i: hidden){
      trans[i]=block1;
    }
    generateM();
  }// add throw later on
}
//public:

HMM::HMM(map<string, map<string ,double>> &trans,
    map<string, map<string ,double>> &emission,
    map<string, double> &state):
  trans{trans},
  emission{emission},
  state{state}
{
  map<string, double> block;
  // generate module
  // put trans keys to hidden
  // put emission subkeys to shown
  generateHS();
  generateM();
}

void HMM::setHiddenToEmit(std::vector<std::vector<std::string>> &v){
  if(this->valid()){
    this->clear();
  }
  if(v.size()>=2){
    hidden=v[0];
    shown=v[1];
  }
  generateTE();
}

void HMM::setTrans(map<string,map<string,double>> &t){
  trans = t;
  generateHS();
  generateM();
}

void HMM::setEmission(map<string, map<string,double>> &e){
  emission = e;
  generateHS();
}

void HMM::setState(map<string, double> &s){
  state = s;
}

double HMM::evaluate(vector<string> &obs)
{
  vector<map<string,double>> states;
  states.emplace_back(state);
  int index = 0;
  double prob = 1.0;
  forwardAlg(obs,states,prob,index);
  return prob;
}

double HMM::evaluate(vector<string> &obs, vector<string> &target){
  vector<map<string,double>> states;
  states.emplace_back(state);
  int index = 0;
  double prob = 1.0;
  forwardAlg(obs,states,prob,index);
  if(target.size()){
    double store = prob;
    forwardAlg(target, states, prob, index);
    prob = prob/store;
  }
  return prob;
}

void HMM::decode(vector<string> &v, vector<string> &store){
  vector<map<string,double>> states;
  for(auto i: trans){
    cout<<i.first<<":"<<endl<<"    ";
    for(auto j: i.second){
      cout<<j.first<<": "<<j.second<<"   ";
    }
    cout<<endl;
  }
  states.emplace_back(state);
  int index = 0;
  viterbiAlg(v,store,states,index);
}

void HMM::setTrans(string &last, string &cur, double prob){
  cout<<"("<<last<<", "<<cur<<")"<<endl;
  if(trans.size()){
    trans[last][cur]=prob;
  }
}

void HMM::setEmission(string &hide, string &show, double prob){
  if(emission.size()){
    emission[hide][show] = prob;
  }
}

void HMM::setState(string &hide, double prob){
  cout<<"set state"<<endl;
  if(module.size() && !state.size()){
    state=module;
  }
  state[hide]=prob;
}
